### PR TITLE
Port augur logic for plasma exits

### PR DESCRIPTION
**Pull request by atvanguard**

### CODE REVIEW

File: contracts/augur/ZeroXExchange.sol [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-import 'ROOT/libraries/token/IERC1155.sol';
+import './external/IExchange.sol';
+// import 'ROOT/libraries/ReentrancyGuard.sol';
+import './external/IWallet.sol';
+import './libraries/math/SafeMathUint256.sol';
+// import 'ROOT/libraries/token/IERC1155.sol';
 
 
-contract ZeroXExchange is IExchange, ReentrancyGuard {
+contract ZeroXExchange is IExchange /* , ReentrancyGuard */ {
```
Comment (atvanguard) [inline contracts/augur/ZeroXExchange.sol]: Will put `ReentrancyGuard`s later on, wherever required.

File: contracts/augur/ZeroXExchange.sol [hunk 2]
```diff
@@ -107,336 +107,6 @@ contract ZeroXExchange is IExchange, ReentrancyGuard {
...
-        totalFillResults.takerFeePaid = totalFillResults.takerFeePaid.add(singleFillResults.takerFeePaid);
-    }
-
-    /// @dev Fills the input order.
-    ///      Returns false if the transaction would otherwise revert.
-    /// @param order Order struct containing order specifications.
-    /// @param takerAssetFillAmount Desired amount of takerAsset to sell.
-    /// @param signature Proof that order has been created by maker.
-    /// @return Amounts filled and fees paid by maker and taker.
-    function fillOrderNoThrow(
```
Comment (atvanguard) [inline contracts/augur/ZeroXExchange.sol]: `fillOrderNoThrow` eventually call `fillOrderInternal` - Calling `exchange.fillOrderInternal`  directly from `ZeroXTrade.sol`

File: contracts/augur/ZeroXExchange.sol [hunk 3]
```diff
@@ -505,11 +175,11 @@ contract ZeroXExchange is IExchange, ReentrancyGuard {
         );
 
         // Settle order
-        settleOrder(
-            order,
-            takerAddress,
-            fillResults
-        );
+        // settleOrder(
```
Comment (atvanguard) [inline contracts/augur/ZeroXExchange.sol]: Not sure if there is a need to call `SettleOrder` and what exactly it settles.

File: contracts/augur/trading/ZeroXTrade.sol [hunk 4]
```diff
@@ -191,26 +48,31 @@ contract ZeroXTrade is Initializable, IZeroXTrade, IERC1155 {
...
 
-        uint256 _protocolFee = 150000 * tx.gasprice;
+        // uint256 _protocolFee = 150000 * tx.gasprice;
 
         // Do the actual asset exchanges
         for (uint256 i = 0; i < _orders.length && _fillAmountRemaining != 0; i++) {
             IExchange.Order memory _order = _orders[i];
-            validateOrder(_order);
-            IExchange _exchange = getExchangeFromAssetData(_order.makerAssetData);
+            // @discuss not sure what validations are required for orders that were signed for Matic and how to retrieve _exchange
```
Comment (atvanguard) [inline contracts/augur/trading/ZeroXTrade.sol]: Let me know, what validations are required.

File: contracts/augur/trading/ZeroXTrade.sol [hunk 5]
```diff
@@ -191,26 +48,31 @@ contract ZeroXTrade is Initializable, IZeroXTrade, IERC1155 {
...
+            // validateOrder(_order);
+
+            // @discuss How to get _exchange address from makerAssetData for an order that was created on Matic
```
Comment (atvanguard) [inline contracts/augur/trading/ZeroXTrade.sol]: **_Discuss:_** How to get _exchange address from makerAssetData for an order that was created on Matic

One solution is, I can have a mapping on the root chain which maps `maticExchangeAddress => maticExchange contract address on rootchain`.
```
IExchange _exchange = getExchangeFromAssetData(_order.makerAssetData);

function getExchangeFromAssetData(makerAssetData) returns (_rootExchangeContract) {
    exchange := mload(add(_assetData, 160))
    _rootExchangeContract = maticToRootExchangeMapping(exchange);
    require(_rootExchangeContract != address(0));
}
```

File: contracts/augur/trading/FillOrder.sol [hunk 6]
```diff
@@ -118,22 +123,6 @@ library Trade {
         });
     }
 
-    function createOrderDataWithOrderId(StoredContracts memory _storedContracts, bytes32 _orderId) internal view returns (OrderData memory) {
```
Comment (atvanguard) [inline contracts/augur/trading/FillOrder.sol]: Like usual, removing parts that are redundant to the predicate.

File: contracts/augur/trading/FillOrder.sol [hunk 7]
```diff
@@ -512,22 +481,24 @@ contract FillOrder is Initializable, ReentrancyGuard, IFillOrder {
...
+        // if (_totalFees > 0) {
+        //     uint256 _longFees = _totalFees.mul(_tradeData.order.sharePriceLong).div(_tradeData.contracts.market.getNumTicks());
+        //     uint256 _shortFees = _totalFees.sub(_longFees);
+        //     _tradeData.contracts.profitLoss.adjustTraderProfitForFees(_tradeData.contracts.market, _tradeData.getLongShareBuyerDestination(), _tradeData.order.outcome, _longFees);
+        //     _tradeData.contracts.profitLoss.adjustTraderProfitForFees(_tradeData.contracts.market, _tradeData.getShortShareBuyerDestination(), _tradeData.order.outcome, _shortFees);
+        // }
+        // updateProfitLoss(_tradeData, _amountFilled);
+        // if (_tradeData.creator.participantAddress == _tradeData.filler.participantAddress) {
+        //     _tradeData.contracts.profitLoss.recordFrozenFundChange(_tradeData.contracts.universe, _tradeData.contracts.market, _tradeData.creator.participantAddress, _tradeData.order.outcome, -int256(_tokensRefunded));
+        // }
```
Comment (atvanguard) [inline contracts/augur/trading/FillOrder.sol]: What should be done with this chunk?

Comment (joeykrug) [reply contracts/augur/trading/FillOrder.sol]: We need to track profit and loss, not sure best way to do it in this system, should discuss w/ Alex

File: contracts/augur/external/IExchange.sol [hunk 8]
```diff
@@ -53,5 +53,8 @@ contract IExchange {
     /// @param takerAssetFillAmount Desired amount of takerAsset to sell.
     /// @param signature Proof that order has been created by maker.
     /// @return Amounts filled and fees paid by maker and taker.
-    function fillOrderNoThrow(Order memory order, uint256 takerAssetFillAmount, bytes memory signature) public payable returns (FillResults memory fillResults);
-}
\ No newline at end of file
+
+    // Replacing fillOrderNoThrow with what it calls internally - fillOrderInternal
+    // Keeping the "Internal" in the function name, to keep it compatible with the function in the original contract
+    function fillOrderInternal(Order memory order, uint256 takerAssetFillAmount, bytes memory signature) public returns (FillResults memory fillResults);
```
Comment (nuevoalex) [inline contracts/augur/external/IExchange.sol]: Isn't this function going to be private/internal? Why was this required?

File: contracts/augur/trading/IAugurTrading.sol [hunk 9]
```diff
@@ -1,16 +1,6 @@
 pragma solidity 0.5.10;
 
 
-import 'ROOT/reporting/IMarket.sol';
-import 'ROOT/reporting/IUniverse.sol';
-
-
 contract IAugurTrading {
     function lookup(bytes32 _key) public view returns (address);
-    function logProfitLossChanged(IMarket _market, address _account, uint256 _outcome, int256 _netPosition, uint256 _avgPrice, int256 _realizedProfit, int256 _frozenFunds, int256 _realizedCost) public returns (bool);
```
Comment (nuevoalex) [inline contracts/augur/trading/IAugurTrading.sol]: We'll need these logs still.

File: contracts/augur/ZeroXExchange.sol (hunk 2 above)
Comment (nuevoalex) [reply contracts/augur/ZeroXExchange.sol]: In reality we won't actually be able to make such changes. This contract is from 0x. We're currently working on a PR that will encorporate their contracts with minimal changes and we won't be able to diverge from their provided interfaces.

File: contracts/augur/trading/FillOrder.sol (hunk 7 above)
Comment (nuevoalex) [reply contracts/augur/trading/FillOrder.sol]: Why does this have to be removed? I'm not sure why the existing logic can't just be ported presently.

File: contracts/augur/trading/FillOrder.sol [hunk 10]
```diff
@@ -556,55 +527,12 @@ contract FillOrder is Initializable, ReentrancyGuard, IFillOrder {
         return true;
     }
 
-    function logOrderFilled(Trade.Data memory _tradeData, uint256 _price, uint256 _fees, uint256 _amountFilled, bytes32 _tradeGroupId) private returns (bool) {
```
Comment (nuevoalex) [inline contracts/augur/trading/FillOrder.sol]: Why are all of these removed? We'll still need to track logs.

File: contracts/augur/trading/ZeroXTrade.sol [hunk 11]
```diff
@@ -1,178 +1,35 @@
...
+import "../trading/IAugurTrading.sol";
+import '../libraries/Initializable.sol';
+import "../IAugur.sol";
+// import 'ROOT/libraries/token/IERC1155.sol';
+
+
+contract ZeroXTrade is Initializable, IZeroXTrade /*, IERC1155 */ {
     using SafeMathUint256 for uint256;
 
-    bool transferFromAllowed = false;
```
Comment (nuevoalex) [inline contracts/augur/trading/ZeroXTrade.sol]: I'm unclear why any changes happened in this contract at all actually. We need all of this removed logic still on Matic.

File: contracts/augur/trading/ZeroXTrade.sol [hunk 12]
```diff
@@ -1,178 +1,35 @@
...
-    }
-
-    function setApprovalForAll(address operator, bool approved) external {
-        revert("Not supported");
-    }
-
-    function isApprovedForAll(address owner, address operator) external view returns (bool) {
-        return true;
-    }
+    // @Discuss Not sure why zeroXTrade requires another ERC1155. How is it different from ShareToken?
```
Comment (nuevoalex) [inline contracts/augur/trading/ZeroXTrade.sol]: This contract fulfills an ERC1155 interface so that it can be specified as the token being traded in 0x signed orders. The actual trade data is packed into the tokenId provided (contains the marketId, price, outcome, and direction) and the makerTokenAssetData (contains the kycToken and exchange address).

File: contracts/augur/trading/ZeroXTrade.sol (hunk 4 above)
Comment (nuevoalex) [reply contracts/augur/trading/ZeroXTrade.sol]: I don't understand why we removed the existing logic for retrieving the exchange form the asset data. Why would this differ for Matic?

File: contracts/augur/trading/ZeroXTrade.sol (hunk 5 above)
Comment (nuevoalex) [reply contracts/augur/trading/ZeroXTrade.sol]: Why can't we just get it form the assetData as we did before?

File: contracts/augur/trading/ZeroXTrade.sol [hunk 13]
```diff
@@ -191,26 +48,31 @@ contract ZeroXTrade is Initializable, IZeroXTrade, IERC1155 {
...
         payable
         returns (uint256)
     {
         uint256 _fillAmountRemaining = _requestedFillAmount;
 
-        transferFromAllowed = true;
+        // transferFromAllowed = true;
 
-        uint256 _protocolFee = 150000 * tx.gasprice;
+        // uint256 _protocolFee = 150000 * tx.gasprice;
```
Comment (nuevoalex) [inline contracts/augur/trading/ZeroXTrade.sol]: This will be required for the 0x v3 exchange contracts.

File: contracts/augur/trading/ZeroXTrade.sol [hunk 14]
```diff
@@ -191,26 +48,31 @@ contract ZeroXTrade is Initializable, IZeroXTrade, IERC1155 {
         address _affiliateAddress,
         bytes32 _tradeGroupId,
         IExchange.Order[] memory _orders,
-        bytes[] memory _signatures
+        bytes[] memory _signatures,
+        address _taker
```
Comment (nuevoalex) [inline contracts/augur/trading/ZeroXTrade.sol]: Why does this take a `taker` param instead of using msg.sender? This would mean anyone could force someone else into being a taker.

File: contracts/augur/reporting/IMarket.sol [hunk 15]
```diff
@@ -1,50 +1,14 @@
...
-    function getEndTime() public view returns (uint256);
-    function getWinningPayoutDistributionHash() public view returns (bytes32);
-    function getWinningPayoutNumerator(uint256 _outcome) public view returns (uint256);
-    function getWinningReportingParticipant() public view returns (IReportingParticipant);
-    function getReputationToken() public view returns (IV2ReputationToken);
-    function getFinalizationTime() public view returns (uint256);
-    function getInitialReporter() public view returns (IInitialReporter);
-    function getDesignatedReportingEndTime() public view returns (uint256);
-    function getValidityBondAttoCash() public view returns (uint256);
-    function deriveMarketCreatorFeeAmount(uint256 _amount) public view returns (uint256);
```
Comment (nuevoalex) [inline contracts/augur/reporting/IMarket.sol]: We'll still need the fee related data for a market to figure out settlement fees that would be taken

File: contracts/augur/reporting/IMarket.sol [hunk 16]
```diff
@@ -1,50 +1,14 @@
...
     }
 
-    function initialize(IAugur _augur, IUniverse _universe, uint256 _endTime, uint256 _feePerCashInAttoCash, uint256 _affiliateFeeDivisor, address _designatedReporterAddress, address _creator, uint256 _numOutcomes, uint256 _numTicks) public;
-    function derivePayoutDistributionHash(uint256[] memory _payoutNumerators) public view returns (bytes32);
-    function doInitialReport(uint256[] memory _payoutNumerators, string memory _description, uint256 _additionalStake) public returns (bool);
-    function getUniverse() public view returns (IUniverse);
-    function getDisputeWindow() public view returns (IDisputeWindow);
     function getNumberOfOutcomes() public view returns (uint256);
     function getNumTicks() public view returns (uint256);
-    function getMarketCreatorSettlementFeeDivisor() public view returns (uint256);
```
Comment (nuevoalex) [inline contracts/augur/reporting/IMarket.sol]: We'll still need the fee related data for a market to figure out settlement fees that would be taken

File: contracts/augur/reporting/IMarket.sol [hunk 17]
```diff
@@ -1,50 +1,14 @@
...
-    function getForkingMarket() public view returns (IMarket _market);
-    function getEndTime() public view returns (uint256);
-    function getWinningPayoutDistributionHash() public view returns (bytes32);
-    function getWinningPayoutNumerator(uint256 _outcome) public view returns (uint256);
-    function getWinningReportingParticipant() public view returns (IReportingParticipant);
```
Comment (nuevoalex) [inline contracts/augur/reporting/IMarket.sol]: These are needed (or some equivalent is needed) to determine payouts for claiming winnings

File: contracts/augur/reporting/IMarket.sol [hunk 18]
```diff
@@ -1,50 +1,14 @@
...
-    function getReputationToken() public view returns (IV2ReputationToken);
-    function getFinalizationTime() public view returns (uint256);
-    function getInitialReporter() public view returns (IInitialReporter);
-    function getDesignatedReportingEndTime() public view returns (uint256);
-    function getValidityBondAttoCash() public view returns (uint256);
-    function deriveMarketCreatorFeeAmount(uint256 _amount) public view returns (uint256);
-    function recordMarketCreatorFees(uint256 _marketCreatorFees, address _affiliateAddress) public returns (bool);
```
Comment (nuevoalex) [inline contracts/augur/reporting/IMarket.sol]: This is used on settlement to distribute creator and affiliate fees which will still be applied

File: contracts/augur/trading/FillOrder.sol (hunk 7 above)
Comment (nuevoalex) [reply contracts/augur/trading/FillOrder.sol]: AFA exit scenarios are concerned we can't really do anything in the mainnet PL contract to automatically update PL as it wont trust this predicate contract,

We _could_ let users modify their own PL via a TX though I'm not sure thats worth the potential risk of it being messed up.
